interop_task/interop_task.md

```diff
@@ -0,0 +1,123 @@
+namespace cl {
+namespace sycl {
+namespace codeplay {
+
+class handler : public cl::sycl::handler {
+ private:
+  // implementation defined constructor
+  handler(__unspecified__);
+
+ public:
+  /* "Manually" enqueue a kernel */
+  template <typename FunctorT>
+  void interop_task(FunctorT hostFunction);
+};
+}  // namespace codeplay
+}  // namespace sycl
+}  // namespace cl
+```
+
+### codeplay::handler::interop_task
+
+The `interop_task` allow the user to execute a task of the native host.
+Unlike `single_task`, `parallel_for` and `parallel_for_work_group`, the `interop_task` do not enqueue a kernel on the device but allow the user to execute a custom action when the prerequisites are satisfied on the device associate with the queue.
+The functor passed to the `interop_task` takes as input a const reference to a `cl::sycl::codeplay::interop_handle` which can be used to retrieve underlying OpenCL objects relative to the execution of the task.
+
+It is not allowed to allocate new SYCL object inside an `interop_task`.
+It is the user responsibilities to ensure all asynchronous executions using SYCL provided resources finished before returning from the `interop_task`.
+
+## Accessing Underlying OpenCL Object
+
+We introduce the `interop_handle` class which provide access to underlying OpenCL objects during the execution of the `interop_task`.
+
+The interface of the `interop_handle` is defined as follow:
+```cpp
+namespace cl {
+namespace sycl {
+namespace codeplay {
+
+class interop_handle {
+ private:
+  // implementation defined constructor
+  interop_handle(__unspecified__);
+
+ public:
+  /* Return the context */
+  cl_context get_context() const;
+
+  /* Return the device id */
+  cl_device_id get_device() const;
+
+  /* Return the command queue associated with this task */
+  cl_command_queue get_queue() const;
+
+  /*
+    Returns the underlying cl_mem object associated with the accessor
+  */
+  template <typename dataT, int dimensions, access::mode accessmode,
+            access::target accessTarget,
+            access::placeholder isPlaceholder>
+  cl_mem get(const accessor<dataT, dimensions, accessmode, access::target accessTarget, access::placeholder isPlaceholder>&) const;
+};
+}  // namespace codeplay
+}  // namespace sycl
+}  // namespace cl
+```
+
+`interop_handle` objects are immutable object whose purpose is to allow the user to access objects relevant to the context.
+
+## Example using regular accessor
+
+```cpp
+    auto cgH = [=] (codeplay::handler& cgh) {
+      auto accA = bufA.get_access<access::mode::read>(cgh);  // Get device accessor to SYCL buffer (cannot be dereference directly in interop_task).
+      auto accB = bufB.get_access<access::mode::read_write>(cgh);
+
+      h.interop_task([=](codeplay::interop_handle &handle) {
+        third_party_api(handle.get_queue(), // Get the OpenCL command queue to use, can be the fallback
+                        handle.get(accA), // Get the OpenCL mem object behind accA
```

Comment:
I am not keen on allowing device accessors in an `interop_task` as there's no way to prevent users from accessing the data via `operator[]` which would modify data without the runtime being aware of it, leading to undefined behaviour.

Reply:
I think having a new type that is constructed in the command group scope, which performs the parameter marshalling in the interop case would provide the functionality required for interop tasks whilst not allowing users to access the data on the host. Having this be a separate type also makes it easier for compilers or static analysis tools to catch undefined behaviour.

Reply:
I agree with you, @Ruyk pointed out that may force users to modify existing code. Going with the accessor route means less code modification is required for users (they can reuse their data types).

That's a trade-off to find, leaving as it is for now, be revisited later.